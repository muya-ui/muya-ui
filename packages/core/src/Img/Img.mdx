---
name: Img 图片
menu: 组件
route: /components/img
autoDemos: true
parent: 数据展示
---

import { LazyDemo } from './demos/styled';

# Img 图片

图片基础功能包含：

- 适配阿里云、金山云自动获取图片后缀
  - 阿里云：[OSS 图片处理文档](https://help.aliyun.com/document_detail/44686.html)
  - 金山云：[图片处理](https://docs.ksyun.com/documents/886)
- 图片懒加载

## 使用场景

在保留原生 img 的特性下，支持懒加载，自定义占位、加载失败等。

## 组件

- `Img`: 默认图片组件，别名 `ImgDiv` 、 `Picture`。
- `ImgSpan`: 设置 `component` = `span`。
- `ImgImg`: 设置 `component` = `img`。
- `ImgContainer`: 图片容器。

{Demos}

## 懒加载逻辑说明

<LazyDemo />

### API

#### ImgProps

<Props propsName="IImgProps" />

#### IImgEvent

图片 `onLoaded` 、 `onError` 的参数。

<Props propsName="IImgEvent" />

#### IImgContainerProps

<Props propsName="IImgContainerProps" />

#### IImgPoolSettings

<Props propsName="IImgPoolSettings" />

### Styles

给不同状态的图片增加不同的 `className`

<Styles typeName="IImgStyleKeys" componentName="img" />
