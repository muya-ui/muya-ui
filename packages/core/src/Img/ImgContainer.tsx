import React, { useContext, useEffect } from 'react';
import styled from 'styled-components';

import memoForwardRef from '../utils/memoForwardRef';
import ImgPoolContext, { ImgPoolProvider } from './ImgPoolContext';
import { IImgContainerProps } from './types';

const BaseNode = styled.div``;
const Container = memoForwardRef<HTMLDivElement, any>((props, ref) => {
  const { children, onScroll, ...otherProps } = props;
  const pool = useContext(ImgPoolContext);
  const innerOnScroll = (e: React.UIEvent<HTMLDivElement>) => {
    onScroll && onScroll(e);
    pool.throttleCheck({ type: 'scroll' });
  };
  useEffect(() => {
    pool.throttleCheck({ type: 'scroll' });
  }, [children, pool]);
  return (
    <BaseNode {...otherProps} onScroll={innerOnScroll} ref={ref}>
      {children}
    </BaseNode>
  );
});

export default memoForwardRef<HTMLDivElement, IImgContainerProps>((props, ref) => {
  const { settings, poolName = 'img_container', ...otherProps } = props;
  return (
    <ImgPoolProvider settings={settings} poolName={poolName}>
      <Container {...otherProps} ref={ref} />
    </ImgPoolProvider>
  );
});
