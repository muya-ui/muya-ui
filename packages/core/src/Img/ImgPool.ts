import deepmerge from 'deepmerge';
import { EventEmitter } from 'events';

import {
  checkWebp,
  IImgSuffix,
  imgSuffix,
  IRectIntersectRect,
  multiRun,
  setRectIntersectContainer,
  wait,
} from '@muya-ui/utils';

import { IImgCheckFn, IImgOption, IImgPool, IImgPoolSettings, IPoolCheckEvent } from './types';

const kssPrefix = ['kss.ksyun.com', 'kssws.ks-cdn.com', '-ks3.kujiale.com'];
/**
 * 判断是否是金山云的地址
 * @param url 地址
 */
export function isKss(url: string) {
  let isKss = false;
  kssPrefix.forEach(prefix => {
    if (url.indexOf(prefix) !== -1) {
      isKss = true;
    }
  });
  return isKss;
}

/**
 * clean 图片后缀
 * @param src 图片原始地址
 * @param oss 阿里云 or 金山云
 */
export function cleanSrc(src: string, oss?: 'ali' | 'kss') {
  if (oss === 'kss') {
    return src.replace(/@base@tag.*$/, '');
  }
  return src.replace(/@.*$/, '').replace(/\?x-oss-process.*$/, '');
}

export function extractFromSrc(src: string, regStr: string) {
  const reg = new RegExp(regStr);
  let external = '';
  src.replace(reg, ($0: string) => {
    external = $0;
    return '';
  });
  return external;
}

export const defaultImgPoolSettings: IImgPoolSettings = {
  options: {
    wait: 'on',
    resize: 'off',
    clean: 'auto',
    lazy: 'on',
    oss: 'auto',
    suffix: 'on',
  },
  concurrent: 6,
  throttleTime: 80,
  suffixs: {
    resizeMode: 'fill',
    ratio: window.devicePixelRatio || 1,
  },
  // defaultImg:
  //   '//qhstaticssl.kujiale.com/newt/23/image/png/1565749312620/129F060E504FDCDAEE29088AB181D326.png',
  // defaultImgStyle: {
  //   backgroundRepeat: 'repeat',
  //   backgroundSize: '100px 100px',
  // },
  checkRegion: {
    top: -0.3,
    bottom: 0.3,
    left: -0.5,
    right: 0.5,
  },
  getImgSrc: (src: string, suffixs: IImgSuffix, options: IImgOption) => {
    if (options.suffix === 'off') {
      return src;
    }
    let ossType: 'kss' | 'ali';
    if (options.oss === 'auto') {
      ossType = isKss(src) ? 'kss' : 'ali';
    } else {
      ossType = options.oss!;
    }

    let innerSrc = src;
    if (options.clean !== 'off') {
      innerSrc = cleanSrc(src, ossType);
      if (options.clean !== 'auto' && options.clean) {
        let external = suffixs.external || '';
        external += extractFromSrc(src, options.clean);
        suffixs.external = external;
      }
    }
    const innerSuffix = imgSuffix(suffixs, ossType);
    return `${innerSrc}${innerSuffix}`;
  },
};

export default class ImgPool extends EventEmitter implements IImgPool {
  scrollCheckFns: Set<IImgCheckFn> = new Set();
  resizeCheckFns: Set<IImgCheckFn> = new Set();
  private _poolStatus: 'none' | 'ready' = 'none';
  private _lastCheckTime: number = -1;
  private _isChecking = false;
  private _lastEventTimer: number = -1;
  private _name: string;
  private _settings: IImgPoolSettings = defaultImgPoolSettings;

  constructor(name?: string) {
    super();
    this._name = name || 'default';
    this.setMaxListeners(30);
  }

  async throttleCheck(e: IPoolCheckEvent) {
    // 正在执行也不继续
    if (this._isChecking) {
      return;
    }
    this._isChecking = true;
    const { throttleTime } = this._settings;

    // 第一次触发的时候，delay 一次
    if (this._lastCheckTime < 0) {
      this._lastCheckTime = Date.now();
      await wait.time(throttleTime!);
    }

    const now = Date.now();
    // 如果具体上传检查结束还不到 100 ms 也不执行检查
    if (now - this._lastCheckTime < throttleTime!) {
      this._isChecking = false;
      return;
    }

    await this.check(e).catch(() => {});
    this._isChecking = false;
    this._lastCheckTime = Date.now();
  }

  debounceCheck(e: IPoolCheckEvent) {
    window.clearTimeout(this._lastEventTimer);
    this._lastEventTimer = window.setTimeout(() => {
      this.check(e).catch(() => {});
    }, this._settings.throttleTime);
  }

  async setup(settings?: Partial<IImgPoolSettings>) {
    if (settings) {
      const { getImgSrc, ...oldSettings } = this._settings;
      const { getImgSrc: newGetImgSrc, ...inputSettings } = settings;
      const newSettings: Partial<IImgPoolSettings> = deepmerge(oldSettings, inputSettings);
      newSettings.getImgSrc = newGetImgSrc || getImgSrc;
      this._settings = newSettings as IImgPoolSettings;
    }
    const scrollFn = () => {
      this.throttleCheck({ type: 'scroll' });
    };
    const resizeFn = () => {
      this._resetRectIntersect();
      this.throttleCheck({ type: 'resize' });
      this.debounceCheck({ type: 'resize' });
    };
    window.addEventListener('scroll', scrollFn, true);
    window.addEventListener('resize', resizeFn);
    this._destroyFn = () => {
      window.removeEventListener('scroll', scrollFn, true);
      window.removeEventListener('resize', resizeFn);
    };
    this._resetRectIntersect();
    if (this._settings.maxListeners) {
      this.setMaxListeners(this._settings.maxListeners);
    }

    const { suffixs } = this._settings;
    if (!suffixs.format && suffixs.webp === undefined) {
      suffixs.webp = await checkWebp();
    }
    this._poolStatus = 'ready';
    this.emit('pool_ready');
  }

  async check(e: IPoolCheckEvent) {
    await this.waitReady();
    if (e.type === 'resize') {
      await this._runFns(this.resizeCheckFns);
    } else {
      await this._runFns(this.scrollCheckFns);
    }
  }

  waitReady() {
    if (this._poolStatus !== 'ready') {
      return new Promise(resolve => {
        this.once('pool_ready', resolve);
      });
    }
    return Promise.resolve();
  }

  destroy() {
    this._poolStatus = 'none';
    this._destroyFn();
  }

  rectIntersect(el: HTMLElement) {
    const rect = this.getRect(el);
    const intersect = this._rectIntersect(rect);

    return { rect, intersect };
  }

  getSuffixs(suffixs: IImgSuffix, rect: Pick<ClientRect, 'width' | 'height'>) {
    const defaultSuffixs = this._settings.suffixs!;
    const result = {
      width: suffixs.width || rect.width || defaultSuffixs.width,
      height: suffixs.height || rect.height || defaultSuffixs.height,
      webp: suffixs.webp !== undefined ? suffixs.webp : defaultSuffixs.webp,
      format: suffixs.format || defaultSuffixs.format,
      resizeMode: suffixs.resizeMode || defaultSuffixs.resizeMode,
      ratio: suffixs.ratio || defaultSuffixs.ratio,
      blur: suffixs.blur || defaultSuffixs.blur,
      q: suffixs.q || defaultSuffixs.q,
      Q: suffixs.Q || defaultSuffixs.Q,
      external: suffixs.external || defaultSuffixs.external,
    };
    return result;
  }

  reset(settings?: Partial<IImgPoolSettings>) {
    this.destroy();
    return this.setup(settings);
  }

  getImgSrc(src: string, suffixs: IImgSuffix, options: IImgOption) {
    const { getImgSrc, options: poolOpts } = this.settings;
    const innerOpt = {
      clean: options.clean || poolOpts!.clean,
      oss: options.oss || poolOpts!.oss,
      suffix: options.suffix || poolOpts!.suffix,
    };

    return getImgSrc(src, suffixs, innerOpt);
  }

  getRect(el: HTMLElement): ClientRect {
    return el.getBoundingClientRect();
  }

  private _destroyFn = () => {};
  private async _runFns(fnSet: Set<IImgCheckFn>) {
    if (fnSet.size === 0) {
      return;
    }
    const fns = Array.from(fnSet.values());
    await multiRun(
      fns,
      async (fn: IImgCheckFn) => {
        await fn();
      },
      this._settings.concurrent,
    );
  }

  private _resetRectIntersect() {
    const { checkRegion } = this._settings;
    this._rectIntersect = setRectIntersectContainer(checkRegion, {
      width: window.innerWidth,
      height: window.innerHeight,
      top: 0,
      left: 0,
    });
  }

  private _rectIntersect: (rect: IRectIntersectRect) => boolean = () => false;

  get poolStatus() {
    return this._poolStatus;
  }

  get settings() {
    return this._settings;
  }

  get name() {
    return this._name;
  }
}
