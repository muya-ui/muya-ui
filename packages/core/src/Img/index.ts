export { default } from './Img';
export { default as Img } from './Img';
export { default as ImgPool } from './ImgPool';
export { default as ImgPoolContext } from './ImgPoolContext';
export { default as ImgContainer } from './ImgContainer';
export { ImgPoolProvider } from './ImgPoolContext';
export { default as useImg } from './useImg';
export * from './Img';
export * from './types';
