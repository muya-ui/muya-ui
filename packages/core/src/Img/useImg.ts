import { useContext, useEffect, useRef, useState } from 'react';
import { shallowEqualObjects } from 'shallow-equal';

import { useMountedRef, wait as utilsWait } from '@muya-ui/utils';

import ImgPoolContext from './ImgPoolContext';
import { IImgHookArgs } from './innerTypes';
import { IImgEvent, IImgLoadStatus, IImgPool } from './types';

type IDiff = Pick<IImgHookArgs, 'suffixs' | 'options' | 'src'>;
export default function useImg(args: IImgHookArgs, pool?: IImgPool) {
  const {
    src,
    imgRef,
    onLoaded,
    onError,
    options: innerOpts,
    suffixs: innerSuffixs,
    loadingDelay,
  } = args;

  const imgPool = useContext(ImgPoolContext);
  const innerPool = pool || imgPool;

  const poolOpts = innerPool.settings.options;
  const innerResize = innerOpts.resize || poolOpts.resize;

  // 状态部分
  const [imgState, setImgState] = useState<IImgEvent>({
    imgSrc: '',
    originSrc: '',
    loadStatus: 'none',
  });

  const lastProps = useRef<IDiff>();
  const mounted = useMountedRef();
  // 当图片已经加载过
  if (lastProps.current && imgState.loadStatus !== 'none') {
    const lastP = lastProps.current;
    // 同时属性发生了变化
    // 再次执行加载逻辑
    if (
      lastP.src !== src ||
      !shallowEqualObjects(innerSuffixs, lastP.suffixs) ||
      !shallowEqualObjects(innerOpts, lastP.options)
    ) {
      setImgState({
        ...imgState,
        loadStatus: 'none',
      });
    }
  }

  function setState(state: IImgEvent) {
    if (!mounted.current) {
      return;
    }
    setImgState(state);
    lastProps.current = {
      src,
      suffixs: innerSuffixs,
      options: innerOpts,
    };
    if (onLoaded && state.loadStatus === 'loaded') {
      onLoaded(state);
      innerPool.emit('img_loaded', state);
    } else if (onError) {
      onError(state);
      innerPool.emit('img_error', state);
    }
  }

  function rectIntersect() {
    if (imgRef.current) {
      return innerPool.rectIntersect(imgRef.current);
    }
    const fakeRect: ClientRect = { top: 0, bottom: 0, left: 0, right: 0, width: 0, height: 0 };
    return { intersect: false, rect: fakeRect };
  }

  const poolReadyPromise = useRef<Promise<void>>();
  const destroyPoolReady = useRef<() => void>(() => {});
  function waitPoolReady() {
    if (innerPool.poolStatus === 'ready') {
      return Promise.resolve();
    }

    if (poolReadyPromise.current) {
      return poolReadyPromise.current;
    }

    poolReadyPromise.current = new Promise(resolve => {
      innerPool.once('pool_ready', resolve);
      destroyPoolReady.current = () => {
        innerPool.removeListener('pool_ready', resolve);
      };
    });

    return poolReadyPromise.current;
  }

  /**
   * 加载图片
   * @param rect
   */
  async function loadImg(rect: ClientRect) {
    await waitPoolReady();
    if (!src) {
      setState({
        originSrc: '',
        loadStatus: 'blank',
        imgSrc: '',
      });
      return;
    }
    const suffixs = innerPool.getSuffixs(innerSuffixs, rect);
    const innerWait = innerOpts.wait || poolOpts.wait;
    const { clean, oss, suffix } = innerOpts;
    let innerImgSrc = innerPool.getImgSrc(src, suffixs, { clean, oss, suffix });
    let loadStatus: IImgLoadStatus = 'loaded';
    let imgInstance;
    if (loadingDelay && loadingDelay > 0) {
      await utilsWait.time(loadingDelay);
    }
    if (innerWait === 'on') {
      try {
        imgInstance = await utilsWait.imgLoaded(innerImgSrc);
      } catch (error) {
        loadStatus = 'error';
      }
    }

    setState({
      originSrc: src,
      loadStatus,
      imgSrc: innerImgSrc,
      rect,
      imgInstance,
    });
  }

  function removeCheckFns() {
    innerPool.scrollCheckFns.delete(scrollCheckFn);
    innerPool.resizeCheckFns.delete(resizeCheckFn);
    destroyPoolReady.current();
  }

  /**
   * 当容器滚动时，检测图片是否需要加载
   */
  async function scrollCheckFn() {
    const { rect, intersect } = rectIntersect();
    if (intersect) {
      removeCheckFns();
      await loadImg(rect);
    }
  }

  /**
   * 当容器缩放时，检测图片是否需要加载
   */
  async function resizeCheckFn() {
    const { rect, intersect } = rectIntersect();
    const { loadStatus } = imgState;

    if (intersect && loadStatus === 'none') {
      removeCheckFns();
      await loadImg(rect);
      return;
    }
    if (loadStatus !== 'loaded') {
      return;
    }

    // 矩形区域是否发生了变更
    const lastRect = imgState.rect;
    const rectChanged =
      lastRect && (rect.width !== lastRect.width || rect.height !== lastRect.height);

    // 得利于 innerPool 一个组件的实例是无法在一次 resize 的检测结束，再就进入下一次检测的
    // 所以这里是不会出现并行两次的情况的
    if (intersect && rectChanged) {
      await loadImg(rect);
    }
    // 不是检测区域内，放到滚动检测，进行滚动检测
    else if (rectChanged) {
      innerPool.scrollCheckFns.add(scrollCheckFn);
    }
  }

  function loadImgDirectly() {
    if (imgRef.current) {
      loadImg(innerPool.getRect(imgRef.current));
    }
  }

  async function checkLoadImg() {
    await waitPoolReady();
    // await utilsWait.time(10000);
    // 关闭了 lazy 则变更就加载
    const innerLazy = innerOpts.lazy || poolOpts.lazy;

    if (innerLazy === 'off') {
      // 中间有几次会触发 unMount
      loadImgDirectly();
      return;
    }
    const { rect, intersect } = rectIntersect();
    if (intersect) {
      loadImg(rect);
      return;
    }
    innerPool.scrollCheckFns.add(scrollCheckFn);
    innerPool.resizeCheckFns.add(resizeCheckFn);
  }

  useEffect(() => {
    if (imgState.loadStatus === 'none') {
      checkLoadImg();
    }
    if (innerResize === 'on') {
      innerPool.resizeCheckFns.add(resizeCheckFn);
    }
    return removeCheckFns;
  });

  return {
    imgPool: innerPool,
    imgState,
    setImgState,
    loadImg,
    scrollCheckFn,
    resizeCheckFn,
    checkLoadImg,
    waitPoolReady,
  };
}
