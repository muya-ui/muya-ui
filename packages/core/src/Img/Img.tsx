import React, { useMemo, useRef } from 'react';
import styled from 'styled-components';

import { useForkRef } from '@muya-ui/utils';

import SpinIcon from '../Spin/SpinIcon';
import { ICustomStyleItem } from '../types';
import memoForwardRef from '../utils/memoForwardRef';
import mergeStyle from '../utils/mergeStyle';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { BaseDiv, BaseImg, BaseSpan, StyledSkeleton } from './styled';
import { IImgImgProps, IImgNode, IImgProps, IImgSpanProps } from './types';
import useImg from './useImg';

const defaultStyles = {
  loading: '',
  error: '',
  blank: '',
  loaded: '',
};
const ImgPure = memoForwardRef<IImgNode, IImgProps>((props, ref) => {
  const {
    // IOSSImgSuffix
    width,
    height,
    suffixHeight,
    suffixWidth,
    webp,
    format,
    resizeMode,
    ratio,
    blur,
    q,
    Q,
    external,
    // IOSSImgOption
    wait,
    resize,
    clean,
    oss,
    suffix,
    lazy,
    // props
    src,
    aspectRatio,
    onLoaded,
    onError,
    defaultImg,
    component,
    defaultImgStyle,
    loadingType,
    loadingDelay,
    // 其余的
    children,
    style,
    styles,
    className,
    ...otherProps
  } = props;
  const imgRef = useRef<IImgNode>(null);
  const handleRef = useForkRef<IImgNode>(imgRef, ref);
  const theme = useTheme();
  const { imgState, imgPool } = useImg({
    imgRef,
    suffixs: {
      width: suffixWidth,
      height: suffixHeight,
      webp,
      format,
      resizeMode,
      ratio,
      blur,
      q,
      Q,
      external,
    },
    options: {
      wait,
      resize,
      clean,
      oss,
      suffix,
      lazy,
    },
    src,
    onError,
    onLoaded,
    loadingDelay,
  });
  const { imgSrc, loadStatus } = imgState;
  const innerStyles = useStyles('img', defaultStyles, styles);
  const loading = loadStatus === 'none';
  const finalDefaultImg =
    defaultImg || imgPool.settings.defaultImg || theme.components.Img.errorImg;

  let useDefaultImgAndIsFromTheme = false;
  let innerImgSrc = imgSrc;
  if (
    (loading && loadingType === 'default-img' && finalDefaultImg) ||
    (loadStatus === 'error' && finalDefaultImg)
  ) {
    innerImgSrc = finalDefaultImg;
    useDefaultImgAndIsFromTheme = finalDefaultImg === theme.components.Img.errorImg;
  }
  let styleItem: ICustomStyleItem | undefined;

  if (loading) {
    styleItem = innerStyles.loading;
  } else if (loadStatus === 'blank') {
    styleItem = innerStyles.blank;
  } else if (loadStatus === 'error') {
    styleItem = innerStyles.error;
  } else if (loadStatus === 'loaded') {
    styleItem = innerStyles.loaded;
  }
  const innerStyle = useMemo(() => {
    let moreStyle;
    if (defaultImg && defaultImgStyle) {
      moreStyle = defaultImgStyle;
    } else if (
      finalDefaultImg === imgPool.settings.defaultImg &&
      imgPool.settings.defaultImgStyle
    ) {
      moreStyle = imgPool.settings.defaultImgStyle;
    }
    const styleItemStyle = styleItem ? styleItem.style : undefined;

    return mergeStyle(moreStyle, styleItemStyle, style);
  }, [
    defaultImg,
    defaultImgStyle,
    finalDefaultImg,
    imgPool.settings.defaultImg,
    imgPool.settings.defaultImgStyle,
    style,
    styleItem,
  ]);

  const imgStyle = useMemo(() => {
    if (innerImgSrc) {
      return {
        ...innerStyle,
        backgroundImage: `url('${innerImgSrc}')`,
      };
    }
    return innerStyle;
  }, [innerImgSrc, innerStyle]);

  const innerClassName = useMemo(() => {
    const classnames: string[] = [];
    if (className) {
      classnames.push(className);
    }
    if (styleItem) {
      classnames.push(styleItem.className);
    }

    return classnames.length > 0 ? classnames.join(' ') : undefined;
  }, [className, styleItem]);

  const finalChildren = useMemo(() => {
    if (loading && loadingType === 'skeleton') {
      return <StyledSkeleton>{children}</StyledSkeleton>;
    } else if (loading && loadingType === 'spin') {
      return <SpinIcon color="currentColor" />;
    }

    return children;
  }, [children, loading, loadingType]);

  if (component === 'img') {
    return (
      <BaseImg
        {...otherProps}
        className={innerClassName}
        theme={theme}
        $width={width}
        $height={height}
        style={innerStyle}
        src={innerImgSrc}
        ref={handleRef as (el: HTMLImageElement) => void}
      />
    );
  }

  if (component === 'span') {
    return (
      <BaseSpan
        {...otherProps}
        className={innerClassName}
        theme={theme}
        $width={width}
        $height={height}
        $aspectRatio={aspectRatio}
        $loadingType={loadingType}
        $useDefaultImgAndIsFromTheme={useDefaultImgAndIsFromTheme}
        $status={loadStatus}
        style={imgStyle}
        ref={handleRef}
      >
        {finalChildren}
      </BaseSpan>
    );
  }
  return (
    <BaseDiv
      {...otherProps}
      className={innerClassName}
      theme={theme}
      $width={width}
      $height={height}
      $aspectRatio={aspectRatio}
      $loadingType={loadingType}
      $status={loadStatus}
      $useDefaultImgAndIsFromTheme={useDefaultImgAndIsFromTheme}
      style={imgStyle}
      ref={handleRef as (el: HTMLDivElement) => void}
    >
      {finalChildren}
    </BaseDiv>
  );
});

const ImgNoTheme = styled(ImgPure)``;
const Img = withThemeForStyled(ImgNoTheme);

const ImgDiv = Img;
const Picture = Img;
const ImgSpan = React.forwardRef<HTMLSpanElement, IImgSpanProps>((props, ref) => {
  const { component, ...otherProps } = props;

  return <Img {...otherProps} ref={ref} component="span" />;
});

const ImgImg = React.forwardRef<HTMLImageElement, IImgImgProps>((props, ref) => {
  const { component, ...otherProps } = props;

  return <Img {...otherProps} ref={ref} component="img" />;
});

export default Img;
export { ImgPure, ImgDiv, ImgSpan, Picture, ImgImg };
