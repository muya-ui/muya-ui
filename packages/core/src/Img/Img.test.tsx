import * as React from 'react';
import renderer from 'react-test-renderer';

import { ImgDiv, ImgImg, ImgSpan } from './Img';
import useImg from './useImg';

jest.mock('./useImg');

test('测试图片节点 & aspectRatio', () => {
  const useImgMock = useImg as jest.Mock;
  useImgMock.mockReturnValue({
    imgState: {
      imgSrc: 'none',
      loadStatus: 'loaded',
    },
    handleRef: () => {},
    imgPool: {
      settings: {},
    },
  });

  const tree = renderer
    .create(
      <>
        <ImgDiv />
        <ImgSpan />
        <ImgImg />

        <ImgDiv width="30%" />
        <ImgDiv width={30} />
        <ImgDiv height="30%" />
        <ImgDiv height={30} />

        <ImgDiv loadingType="skeleton" />
        <ImgDiv loadingType="spin" />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('aspectRatio', () => {
  const useImgMock = useImg as jest.Mock;
  useImgMock.mockReturnValue({
    imgState: {
      imgSrc: 'none',
      loadStatus: 'loaded',
    },
    handleRef: () => {},
    imgPool: {
      settings: {},
    },
  });

  const tree = renderer
    .create(
      <>
        <ImgDiv aspectRatio="1:1" />
        <ImgDiv aspectRatio="4:3" />
        <ImgDiv aspectRatio="3:2" />
        <ImgDiv aspectRatio="16:9" />
        <ImgDiv aspectRatio="3:4" />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
test('aspectRatio 同时设置宽度为数字', () => {
  const useImgMock = useImg as jest.Mock;
  useImgMock.mockReturnValue({
    imgState: {
      imgSrc: 'none',
      loadStatus: 'loaded',
    },
    handleRef: () => {},
    imgPool: {
      settings: {},
    },
  });

  const tree = renderer
    .create(
      <>
        <ImgDiv aspectRatio="1:1" width={100} />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试图片 error', () => {
  const useImgMock = useImg as jest.Mock;
  useImgMock.mockReturnValue({
    imgState: {
      imgSrc: 'none',
      loadStatus: 'error',
    },
    handleRef: () => {},
    imgPool: {
      settings: {
        defaultImg: 'aa',
      },
    },
  });

  const tree1 = renderer.create(<ImgDiv />).toJSON();
  expect(tree1).toMatchSnapshot();
  const tree2 = renderer.create(<ImgDiv defaultImg="aaa" />).toJSON();
  expect(tree2).toMatchSnapshot();
});
