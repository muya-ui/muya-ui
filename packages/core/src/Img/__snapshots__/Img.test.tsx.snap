// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aspectRatio 1`] = `
Array [
  .c0 {
  position: relative;
  height: 0;
  padding-bottom: 100%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  position: relative;
  height: 0;
  padding-bottom: 75%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  position: relative;
  height: 0;
  padding-bottom: 66.66666666666666%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  position: relative;
  height: 0;
  padding-bottom: 56.25%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  position: relative;
  height: 0;
  padding-bottom: 133.33333333333331%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
]
`;

exports[`aspectRatio 同时设置宽度为数字 1`] = `
.c0 {
  position: relative;
  height: 100px;
  width: 100px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
  className="c0 muya-img-loaded"
  style={
    Object {
      "backgroundImage": "url('none')",
    }
  }
/>
`;

exports[`测试图片 error 1`] = `
.c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
  className="c0 muya-img-error"
  style={
    Object {
      "backgroundImage": "url('aa')",
    }
  }
/>
`;

exports[`测试图片 error 2`] = `
.c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
  className="c0 muya-img-error"
  style={
    Object {
      "backgroundImage": "url('aaa')",
    }
  }
/>
`;

exports[`测试图片节点 & aspectRatio 1`] = `
Array [
  .c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  display: inline-block;
}

<span
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  <img
    className="muya-img-loaded"
    src="none"
  />,
  .c0 {
  width: 30%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  width: 30px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  height: 30%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  height: 30px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
  .c0 {
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  -webkit-transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
  transition: background-image 300ms cubic-bezier(0.0,0,0.2,1);
}

<div
    className="c0 muya-img-loaded"
    style={
      Object {
        "backgroundImage": "url('none')",
      }
    }
  />,
]
`;
